apps/designer/app/canvas/features/text-editor/interop.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  type ElementNode,
+  $getRoot,
+  $createTextNode,
+  $createParagraphNode,
+  TextNode,
+  LineBreakNode,
+  ParagraphNode,
+} from "lexical";
+import { $createLinkNode, LinkNode } from "@lexical/link";
+import type { ChildrenUpdates, Instance } from "@webstudio-is/react-sdk";
+import { createInstanceId } from "~/shared/tree-utils";
+
+// Map<nodeKey, Instance>
+export type Refs = Map<string, Instance>;
+
+const $writeUpdates = (
+  node: ElementNode,
+  updates: ChildrenUpdates,
+  refs: Refs
+) => {
+  const children = node.getChildren();
+  for (const child of children) {
+    if (child instanceof ParagraphNode) {
+      if (updates.length !== 0) {
+        updates.push("\n");
+      }
+      $writeUpdates(child, updates, refs);
+    }
+    if (child instanceof LineBreakNode) {
+      if (updates.length !== 0) {
+        // @todo should we visually distinct line breaks and paragraphs?
```

Comment:
can you explain what the question is? I it the difference between creating new paragraphs when hitting enter? vs using the same paragraph with newlines inside?

Reply:
If use default browser styling paragraphs have not editable space (margin) between them, while line break is just a newline. Lexical allows to enter line break with shift+enter.